src/Tests/Microsoft.Diagnostics.Monitoring.UnitTests/AuthenticationTests.cs

```diff
@@ -0,0 +1,232 @@
+
+            _outputHelper.WriteLine("Rotating API key.");
+
+            // Rotate the API key
+            byte[] apiKey2 = GenerateApiKey();
+
+            options.UseApiKey(AlgorithmName, apiKey2);
+            toolRunner.WriteKeyPerValueConfiguration(options);
+
+            // Wait for the key rotation to be consumed by dotnet-monitor; detect this
+            // by checking for when API returns a 401. Ideally, key rotation would write
+            // log event and runner monitor for event and notify.
+            int attempts = 0;
+            while (true)
+            {
+                attempts++;
+                _outputHelper.WriteLine("Waiting for key rotation (attempt #{0}).", attempts);
+
+                try
+                {
+                    await client.GetProcessesAsync(DefaultApiTimeout);
+                }
+                catch (ApiStatusCodeException ex) when (ex.StatusCode == HttpStatusCode.Unauthorized)
+                {
+                    break;
+                }
+
+                await Task.Delay(TimeSpan.FromSeconds(3));
+
+                Assert.True(attempts < 10);
+            }
+
+            _outputHelper.WriteLine("Verifying new API key.");
+
+            // Use new API key
+            httpClient.DefaultRequestHeaders.Authorization =
+                new AuthenticationHeaderValue(ApiKeyScheme, Convert.ToBase64String(apiKey2));
+
+            // Check that /processes does not challenge for authentication
+            processes = await client.GetProcessesAsync(DefaultApiTimeout);
+            Assert.NotNull(processes);
+        }
+
+        /// <summary>
+        /// Tests that Negotiate authentication can be used for authentication.
+        /// </summary>
+        [Fact]
+        public async Task NegotiateAuthenticationSchemeTest()
+        {
+            await using DotNetMonitorRunner toolRunner = new DotNetMonitorRunner(_outputHelper);
+            await toolRunner.StartAsync(DefaultStartTimeout);
+
+            // Create HttpClient and HttpClientHandler that uses the current
+            // user's credentials from the test process. Since dotnet-monitor
+            // is launched by the test process, the usage of these credentials
+            // should authenticate correctly (except when elevated, which the
+            // tool will deny authorization).
+            using HttpClientHandler handler = new HttpClientHandler();
+            handler.Credentials = CredentialCache.DefaultCredentials;
+            using HttpClient httpClient = new HttpClient(handler);
+
+            using ApiClient client = new ApiClient(_outputHelper, await toolRunner.GetDefaultAddressAsync(DefaultAddressTimeout), httpClient);
+
+            // TODO: Split test into elevated vs non-elevated tests and skip
+            // when not running in the corresponding context. Possibly unelevate
+            // dotnet-monitor when running tests elevated.
+            if (EnvironmentInformation.IsElevated)
```

Comment:
Do we even have a way of running this elevated during PR checks?

Reply:
Fun thing I learned today: the CI build runs them elevated. So what I really need to do if figure out a way to de-elevate. I think I found some Win32 APIs for creating an access token with normal privileges and then can use CreateProcessAsUser to launch the executables, so I can rewrite the DotNetRunner to de-elevate (maybe by default).

Reply:
I don't plan on making the de-elevation change in this PR.